src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,528 @@
+						ex);
+				}
+			}
+
+			/// <summary>
+			/// Lock the context, causing <see cref="Wait"/> to block until released. Do nothing if the context
+			/// has already been locked once.
+			/// </summary>
+			protected virtual void Lock()
+			{
+				if (_locked || _isDisposed)
+					return;
+				_locked = true;
+				_semaphore = new SemaphoreSlim(0);
+			}
+
+			/// <summary>
+			/// Unlock the context, causing <see cref="Wait"/> to cease blocking. Do nothing if the context
+			/// is not locked.
+			/// </summary>
+			protected virtual void Unlock()
+			{
+				// Do not set _locked back to false, we want to wait only once per transaction context.
+				// Volatile member, do not simplify by removing the local variable.
+				var semaphore = _semaphore;
+				_semaphore = null;
+				if (semaphore == null)
+					return;
+				// Not supposed to have more than one waiting thread, but in case the user made bad multi-threading,
+				// do not leave frozen threads.
+				semaphore.Release(100);
+				semaphore.Dispose();
+			}
+
+			#region IEnlistmentNotification Members
+
+			/// <summary>
+			/// Prepare the session for the transaction commit. Runs
+			/// <see cref="ISessionImplementor.BeforeTransactionCompletion(ITransaction)"/> for the session and for
+			/// <see cref="ConnectionManager.DependentSessions"/> if any. <see cref="Lock"/> the context
+			/// before signaling it is done, or before rollback in case of failure.
+			/// </summary>
+			/// <param name="preparingEnlistment">The object for notifying the prepare phase outcome.</param>
+			public virtual void Prepare(PreparingEnlistment preparingEnlistment)
+			{
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+				{
+					try
+					{
+						using (_sessionImplementor.ConnectionManager.BeginsProcessingFromSystemTransaction(_useConnectionOnSystemTransactionEvents))
+						{
+							if (_useConnectionOnSystemTransactionEvents)
+							{
+								// Ensure any newly acquired connection gets enlisted in the transaction. When distributed,
+								// this code runs from another thread and we cannot rely on Transaction.Current.
+								using (var tx = new TransactionScope(AmbientTransaction))
+								{
+									_sessionImplementor.BeforeTransactionCompletion(null);
+									foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+										dependentSession.BeforeTransactionCompletion(null);
+
+									tx.Complete();
+								}
+							}
+							else
+							{
+								_sessionImplementor.BeforeTransactionCompletion(null);
+								foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+									dependentSession.BeforeTransactionCompletion(null);
+							}
+						}
+						// Lock the session to ensure second phase gets done before the session is used by code following
+						// the transaction scope disposal.
+						Lock();
```

Comment:
It must be called before `preparingEnlistment.Done()` or before `ForceRollback`. As soon as those methods are called, MSDTC may trigger the second phase of the 2PC and the completion event on others threads, without waiting the remaining of the callback execution. Tests in `ResourceManagerFixture` show that.

So calling it in a finally would be a bit impractical.